In [20]:
import pandas as pd
import re
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# **Telecommunication company customer churn prediction using different machine leargning algorithms for classifying**

## **ABSTRACT**
In th

## **1.Introduction**

Customer churn, defined as the act of customers discontinuing their service [1], is a significant challenge faced by the telecommunications industry, where retaining existing customers is often more cost-effective than acquiring new ones [2]. Churn refers to the number of customers who leave a service provider over a certain period [3], and predicting which customers are likely to churn is critical to minimizing revenue losses [4]. The goal of churn prediction is to identify potential churners early, so that businesses can take proactive steps, such as offering personalized incentives and retention strategies, to keep them.

Telecom companies face high acquisition costs compared to retention costs [2], and predicting churn accurately can prevent the misallocation of resources. When the prediction of churn is incorrect, companies may waste money on customers who would not have left, while missing opportunities to retain those who were actually at risk. Studies show that acquisition costs can be up to five times higher than retention costs [5], making it crucial to target the right customers with retention efforts.

With the rise of popularity of data-driven solutions, machine learning has emerged as a powerful tool to predict customer churn [6] based on historical data. By analyzing patterns in customer behavior, service usage, billing information, and other relevant factors like demographic features and more, machine learning models can provide insights that were previously difficult to capture through traditional methods. These predictive models allow businesses to effectively target at-risk customers, offering retention campaigns before the churn occurs [7].Using algorithms such as Random Forests, Support Vector Machines (SVM), Decision Trees and more, companies could gain valuable insights into customer behavior and identify the key factors that drive churn. Accurate churn prediction models can enable telecom companies to reduce churn rates, increase customer loyalty, and enhance profitability. [8]

In addition to improving churn prediction, feature selection and the identification of relevant factors that influence churn may be critical for building effective models. [9] Telecom customers churn for a variety of reasons, and these factors should be considered to develop personalized retention strategies. Machine learning models allow for a more realistic and targeted approach, splitting customers based on their churn risk and offering the companies to develop customized solutions to prevent them from leaving.

This project aims to explore the factors that lead to customer churn in order to allocate resources on their improvement or to gain knowledge on where the resources could be spared and to develop a churn prediction model using machine learning algorithms, with the goal of helping telecom companies predict customer churn and improve customer retention strategies. For that we will be trying different ML algorithms in order to find and tune the best model for our goal - to find out which customers would churn so that resources and strategies are built for their retention and which wouldn't so that resources aren't misallocated into retaining a non churner.

## **2. Related work**

1. [9] Ullah et al. (2019) - This paper used data preprocessing with noise removal and feature selection, applied multiple classifiers including(Random Forest, Decision Tree, Naive Bayes, Multilayer Perceptron (MLP), and Logistic Regression.) employed k-means clustering for customer segmentation, used rule-based algorithms for churn factor identification, and evaluated the model with standard metrics like accuracy and ROC area.

2. [10] Wagh et al. (2023) - This paper used data preprocessing to clean data, Pearson correlation for feature selection, applied Random Forest and Decision Tree for classification, utilized Cox Proportional Hazard Model and Kaplan-Meier analysis for survival prediction, and used SMOTE and ENN for handling imbalanced datasets.

3. [11] Ahmed and Maheswari (2017) - The paper used a hybrid Firefly algorithm, enhanced with Simulated Annealing to optimize comparisons between fireflies, for more efficient classification on large and sparse telecom datasets, and evaluated the model using metrics like ROC, Precision-Recall, F-Measure, and accuracy

4. [12] Jain et al. (2020) - The paper used Logistic Regression and Logit Boost machine learning algorithms, applied after data cleaning and preparation, to predict customer churn, and evaluated their performance using metrics such as Kappa statistic, Mean Absolute Error, Root Mean Square Error, and Accuracy

5. [13] Poudel et al. (2024) - The paper used data preprocessing to clean telecom data, implemented feature engineering and one-hot encoding for categorical variables, applied various machine learning models like SVM, Logistic Regression, Random Forest, Gradient Boosting Machine (GBM), and Neural Networks for churn prediction, evaluated models using 10-fold cross-validation, and employed SHAP (SHapley Additive Explanations) plots for global and local interpretability of predictions​

## **3.Data acquisition and cleaning**

### **Data acquisition**

The data is acquired through Kaggle and it's free for use. This dataset contains detailed information about customers of a telecom company, including demographics, service subscriptions, billing information, and churn status. It captures key aspects such as whether a customer has phone or internet services, their tenure with the company, usage of additional services like online security, and their chosen payment methods. The dataset is particularly useful for analyzing patterns and factors that contribute to customer churn, helping the company understand and potentially mitigate reasons for customer departure. [14]

Telecom Customers, By: Turek Muhammed, License: Apache 2.0, Name of file: Telecom Customers Churn.csv   Link: https://www.kaggle.com/datasets/tarekmuhammed/telecom-customers/data

### **Data cleaning**

1. First let's read the dataset using Pandas's read_csv function, see the shape of it (observations and features) and it's columns and potential problems like missing values, duplicates and check the data constistency(whether a numeric feature is indeed in an numeric datatype).

In [24]:
telecom_data = pd.read_csv("Telecom Customers Churn.csv")

2. Let's see the DataFrame and the shape of the data to gain some knowledge on what the data looks like.

In [27]:
telecom_data

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
0     7590-VHVEG  Female              0     Yes         No       1   
1     5575-GNVDE    Male              0      No         No      34   
2     3668-QPYBK    Male              0      No         No       2   
3     7795-CFOCW    Male              0      No         No      45   
4     9237-HQITU  Female              0      No         No       2   
...          ...     ...            ...     ...        ...     ...   
7038  6840-RESVB    Male              0     Yes        Yes      24   
7039  2234-XADUH  Female              0     Yes        Yes      72   
7040  4801-JZAZL  Female              0     Yes        Yes      11   
7041  8361-LTMKD    Male              1     Yes         No       4   
7042  3186-AJIEK    Male              0      No         No      66   

     PhoneService     MultipleLines InternetService OnlineSecurity  ...  \
0              No  No phone service             DSL             No  ...   
1             Yes                No             DSL            Yes  ...   
2             Yes                No             DSL            Yes  ...   
3              No  No phone service             DSL            Yes  ...   
4             Yes                No     Fiber optic             No  ...   
...           ...               ...             ...            ...  ...   
7038          Yes               Yes             DSL            Yes  ...   
7039          Yes               Yes     Fiber optic             No  ...   
7040           No  No phone service             DSL            Yes  ...   
7041          Yes               Yes     Fiber optic             No  ...   
7042          Yes                No     Fiber optic            Yes  ...   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                  No          No          No              No  Month-to-month   
1                 Yes          No          No              No        One year   
2                  No          No          No              No  Month-to-month   
3                 Yes         Yes          No              No        One year   
4                  No          No          No              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7038              Yes         Yes         Yes             Yes        One year   
7039              Yes          No         Yes             Yes        One year   
7040               No          No          No              No  Month-to-month   
7041               No          No          No              No  Month-to-month   
7042              Yes         Yes         Yes             Yes        Two year   

     PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
0                 Yes           Electronic check          29.85         29.85   
1                  No               Mailed check          56.95        1889.5   
2                 Yes               Mailed check          53.85        108.15   
3                  No  Bank transfer (automatic)          42.30       1840.75   
4                 Yes           Electronic check          70.70        151.65   
...               ...                        ...            ...           ...   
7038              Yes               Mailed check          84.80        1990.5   
7039              Yes    Credit card (automatic)         103.20        7362.9   
7040              Yes           Electronic check          29.60        346.45   
7041              Yes               Mailed check          74.40         306.6   
7042              Yes  Bank transfer (automatic)         105.65        6844.5   

     Churn  
0       No  
1       No  
2      Yes  
3       No  
4      Yes  
...    ...  
7038    No  
7039    No  
7040    No  
7041   Yes  
7042    No  

[7043 rows x 21 columns]

In [47]:
telecom_data.shape

(7043, 21)

3. As it can be seen the dataset has 7043 observations(rows) and 21 features(columns), from which the column "Churn" is the target variable.

In [62]:
telecom_data.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

4. The names of the columns could be transformed into snake_case for easier usage. I will create a function to transform the column names into snake_case

In [70]:
def convert_to_snake_case(column_name):
    """
    Convert a string to snake_case.

    Parameters:
    column_name (str): The original column name.

    Returns:
    str: The column name converted to snake_case.
    """
  
    snake_case_name = re.sub(r'[\W\s]+', '_', column_name)
    
   
    snake_case_name = re.sub(r'(?<!^)(?=[A-Z])', '_', snake_case_name)
    

    snake_case_name = snake_case_name.lower()
    
   
    snake_case_name = re.sub(r'_+', '_', snake_case_name)
    
   
    snake_case_name = snake_case_name.strip('_')
    
    return snake_case_name

def rename_columns_to_snake_case(df):
    """
    Rename all columns in a DataFrame to snake_case.

    Parameters:
    df (pd.DataFrame): The DataFrame whose columns need to be renamed.

    Returns:
    pd.DataFrame: A new DataFrame with columns renamed to snake_case.
    """
    
    column_mapping = {col: convert_to_snake_case(col) for col in df.columns}
    
    
    df_snake_case = df.rename(columns=column_mapping)
    
    return df_snake_case

In [72]:
telecom_customer_data_snake_case = rename_columns_to_snake_case(telecom_data)

5. Let's see the result and whether all columns transformed

In [75]:
telecom_customer_data_snake_case

customer_i_d  gender  senior_citizen partner dependents  tenure  \
0      7590-VHVEG  Female               0     Yes         No       1   
1      5575-GNVDE    Male               0      No         No      34   
2      3668-QPYBK    Male               0      No         No       2   
3      7795-CFOCW    Male               0      No         No      45   
4      9237-HQITU  Female               0      No         No       2   
...           ...     ...             ...     ...        ...     ...   
7038   6840-RESVB    Male               0     Yes        Yes      24   
7039   2234-XADUH  Female               0     Yes        Yes      72   
7040   4801-JZAZL  Female               0     Yes        Yes      11   
7041   8361-LTMKD    Male               1     Yes         No       4   
7042   3186-AJIEK    Male               0      No         No      66   

     phone_service    multiple_lines internet_service online_security  ...  \
0               No  No phone service              DSL              No  ...   
1              Yes                No              DSL             Yes  ...   
2              Yes                No              DSL             Yes  ...   
3               No  No phone service              DSL             Yes  ...   
4              Yes                No      Fiber optic              No  ...   
...            ...               ...              ...             ...  ...   
7038           Yes               Yes              DSL             Yes  ...   
7039           Yes               Yes      Fiber optic              No  ...   
7040            No  No phone service              DSL             Yes  ...   
7041           Yes               Yes      Fiber optic              No  ...   
7042           Yes                No      Fiber optic             Yes  ...   

     device_protection tech_support streaming_t_v streaming_movies  \
0                   No           No            No               No   
1                  Yes           No            No               No   
2                   No           No            No               No   
3                  Yes          Yes            No               No   
4                   No           No            No               No   
...                ...          ...           ...              ...   
7038               Yes          Yes           Yes              Yes   
7039               Yes           No           Yes              Yes   
7040                No           No            No               No   
7041                No           No            No               No   
7042               Yes          Yes           Yes              Yes   

            contract paperless_billing             payment_method  \
0     Month-to-month               Yes           Electronic check   
1           One year                No               Mailed check   
2     Month-to-month               Yes               Mailed check   
3           One year                No  Bank transfer (automatic)   
4     Month-to-month               Yes           Electronic check   
...              ...               ...                        ...   
7038        One year               Yes               Mailed check   
7039        One year               Yes    Credit card (automatic)   
7040  Month-to-month               Yes           Electronic check   
7041  Month-to-month               Yes               Mailed check   
7042        Two year               Yes  Bank transfer (automatic)   

     monthly_charges  total_charges churn  
0              29.85          29.85    No  
1              56.95         1889.5    No  
2              53.85         108.15   Yes  
3              42.30        1840.75    No  
4              70.70         151.65   Yes  
...              ...            ...   ...  
7038           84.80         1990.5    No  
7039          103.20         7362.9    No  
7040           29.60         346.45    No  
7041           74.40          306.6   Yes  
7042          105.65         6844.5    No  

[704

In [77]:
telecom_customer_data_snake_case.columns

Index(['customer_i_d', 'gender', 'senior_citizen', 'partner', 'dependents',
       'tenure', 'phone_service', 'multiple_lines', 'internet_service',
       'online_security', 'online_backup', 'device_protection', 'tech_support',
       'streaming_t_v', 'streaming_movies', 'contract', 'paperless_billing',
       'payment_method', 'monthly_charges', 'total_charges', 'churn'],
      dtype='object')

6. Now that all the columns are renamed let's continue onward with checking for NaN values, duplicates and datatype consistency

In [81]:
telecom_customer_data_snake_case.isnull().sum()

customer_i_d         0
gender               0
senior_citizen       0
partner              0
dependents           0
tenure               0
phone_service        0
multiple_lines       0
internet_service     0
online_security      0
online_backup        0
device_protection    0
tech_support         0
streaming_t_v        0
streaming_movies     0
contract             0
paperless_billing    0
payment_method       0
monthly_charges      0
total_charges        0
churn                0
dtype: int64

7. Apparently no NaN values are found, let's check for duplicates.

In [88]:
telecom_customer_data_snake_case.duplicated().sum()

0

8. No duplicates found aswell, let's now see if the data types are right

In [92]:
telecom_customer_data_snake_case.dtypes

customer_i_d          object
gender                object
senior_citizen         int64
partner               object
dependents            object
tenure                 int64
phone_service         object
multiple_lines        object
internet_service      object
online_security       object
online_backup         object
device_protection     object
tech_support          object
streaming_t_v         object
streaming_movies      object
contract              object
paperless_billing     object
payment_method        object
monthly_charges      float64
total_charges         object
churn                 object
dtype: object

9. All clooks okay except for total_charges which is of type object which makes me think there is a null value or something that is not encoded right so I will look into it
10. I will create a function to display the unique values and their count in every column

In [102]:
def display_unique_values_and_counts(df):
    """
    Display unique values and their counts for each column in a DataFrame.
    
    Parameters:
    df (pd.DataFrame): The DataFrame whose columns will be analyzed.
    
    Returns:
    None
    """
    for column in df.columns:
        print(f"Column: {column}")
        print(df[column].value_counts(dropna = False))
        print("-" * 50)  


display_unique_values_and_counts(telecom_customer_data_snake_case)

Column: customer_i_d
customer_i_d
7590-VHVEG    1
3791-LGQCY    1
6008-NAIXK    1
5956-YHHRX    1
5365-LLFYV    1
             ..
9796-MVYXX    1
2637-FKFSY    1
1552-AAGRX    1
4304-TSPVK    1
3186-AJIEK    1
Name: count, Length: 7043, dtype: int64
--------------------------------------------------
Column: gender
gender
Male      3555
Female    3488
Name: count, dtype: int64
--------------------------------------------------
Column: senior_citizen
senior_citizen
0    5901
1    1142
Name: count, dtype: int64
--------------------------------------------------
Column: partner
partner
No     3641
Yes    3402
Name: count, dtype: int64
--------------------------------------------------
Column: dependents
dependents
No     4933
Yes    2110
Name: count, dtype: int64
--------------------------------------------------
Column: tenure
tenure
1     613
72    362
2     238
3     200
4     176
     ... 
28     57
39     56
44     51
36     50
0      11
Name: count, Length: 73, dtype: int64
---------

11. Here we can see that 11 of the total_charge entries are blank that's why it doesn't count them as null, but they arent numbers either, so I shall remove them, but let's make sure 

In [107]:
telecom_customer_data_snake_case['total_charges'] = pd.to_numeric(telecom_customer_data_snake_case['total_charges'], errors='coerce')
total_charges_issues = telecom_customer_data_snake_case['total_charges'].isnull().sum()
print(f"Number of invalid 'TotalCharges' entries: {total_charges_issues}")

Number of invalid 'TotalCharges' entries: 11


12. Let's drop these nan values

In [110]:
telecom_customer_data_cleaned_nan = telecom_customer_data_snake_case.dropna(subset=['total_charges'])

In [114]:
telecom_customer_data_cleaned_nan 

customer_i_d  gender  senior_citizen partner dependents  tenure  \
0      7590-VHVEG  Female               0     Yes         No       1   
1      5575-GNVDE    Male               0      No         No      34   
2      3668-QPYBK    Male               0      No         No       2   
3      7795-CFOCW    Male               0      No         No      45   
4      9237-HQITU  Female               0      No         No       2   
...           ...     ...             ...     ...        ...     ...   
7038   6840-RESVB    Male               0     Yes        Yes      24   
7039   2234-XADUH  Female               0     Yes        Yes      72   
7040   4801-JZAZL  Female               0     Yes        Yes      11   
7041   8361-LTMKD    Male               1     Yes         No       4   
7042   3186-AJIEK    Male               0      No         No      66   

     phone_service    multiple_lines internet_service online_security  ...  \
0               No  No phone service              DSL              No  ...   
1              Yes                No              DSL             Yes  ...   
2              Yes                No              DSL             Yes  ...   
3               No  No phone service              DSL             Yes  ...   
4              Yes                No      Fiber optic              No  ...   
...            ...               ...              ...             ...  ...   
7038           Yes               Yes              DSL             Yes  ...   
7039           Yes               Yes      Fiber optic              No  ...   
7040            No  No phone service              DSL             Yes  ...   
7041           Yes               Yes      Fiber optic              No  ...   
7042           Yes                No      Fiber optic             Yes  ...   

     device_protection tech_support streaming_t_v streaming_movies  \
0                   No           No            No               No   
1                  Yes           No            No               No   
2                   No           No            No               No   
3                  Yes          Yes            No               No   
4                   No           No            No               No   
...                ...          ...           ...              ...   
7038               Yes          Yes           Yes              Yes   
7039               Yes           No           Yes              Yes   
7040                No           No            No               No   
7041                No           No            No               No   
7042               Yes          Yes           Yes              Yes   

            contract paperless_billing             payment_method  \
0     Month-to-month               Yes           Electronic check   
1           One year                No               Mailed check   
2     Month-to-month               Yes               Mailed check   
3           One year                No  Bank transfer (automatic)   
4     Month-to-month               Yes           Electronic check   
...              ...               ...                        ...   
7038        One year               Yes               Mailed check   
7039        One year               Yes    Credit card (automatic)   
7040  Month-to-month               Yes           Electronic check   
7041  Month-to-month               Yes               Mailed check   
7042        Two year               Yes  Bank transfer (automatic)   

     monthly_charges  total_charges  churn  
0              29.85          29.85     No  
1              56.95        1889.50     No  
2              53.85         108.15    Yes  
3              42.30        1840.75     No  
4              70.70         151.65    Yes  
...              ...            ...    ...  
7038           84.80        1990.50     No  
7039          103.20        7362.90     No  
7040           29.60         346.45     No  
7041           74.40         306.60    Yes  
7042          105.65        6844.50   

12. Let's remove the customer_i_d column as it doesn't provide any valuable information 

In [128]:
telecom_data_cleaned = telecom_customer_data_cleaned_nan.drop(['customer_i_d'], axis=1,).reset_index(drop=True)

In [130]:
telecom_data_cleaned

gender  senior_citizen partner dependents  tenure phone_service  \
0     Female               0     Yes         No       1            No   
1       Male               0      No         No      34           Yes   
2       Male               0      No         No       2           Yes   
3       Male               0      No         No      45            No   
4     Female               0      No         No       2           Yes   
...      ...             ...     ...        ...     ...           ...   
7027    Male               0     Yes        Yes      24           Yes   
7028  Female               0     Yes        Yes      72           Yes   
7029  Female               0     Yes        Yes      11            No   
7030    Male               1     Yes         No       4           Yes   
7031    Male               0      No         No      66           Yes   

        multiple_lines internet_service online_security online_backup  \
0     No phone service              DSL              No           Yes   
1                   No              DSL             Yes            No   
2                   No              DSL             Yes           Yes   
3     No phone service              DSL             Yes            No   
4                   No      Fiber optic              No            No   
...                ...              ...             ...           ...   
7027               Yes              DSL             Yes            No   
7028               Yes      Fiber optic              No           Yes   
7029  No phone service              DSL             Yes            No   
7030               Yes      Fiber optic              No            No   
7031                No      Fiber optic             Yes            No   

     device_protection tech_support streaming_t_v streaming_movies  \
0                   No           No            No               No   
1                  Yes           No            No               No   
2                   No           No            No               No   
3                  Yes          Yes            No               No   
4                   No           No            No               No   
...                ...          ...           ...              ...   
7027               Yes          Yes           Yes              Yes   
7028               Yes           No           Yes              Yes   
7029                No           No            No               No   
7030                No           No            No               No   
7031               Yes          Yes           Yes              Yes   

            contract paperless_billing             payment_method  \
0     Month-to-month               Yes           Electronic check   
1           One year                No               Mailed check   
2     Month-to-month               Yes               Mailed check   
3           One year                No  Bank transfer (automatic)   
4     Month-to-month               Yes           Electronic check   
...              ...               ...                        ...   
7027        One year               Yes               Mailed check   
7028        One year               Yes    Credit card (automatic)   
7029  Month-to-month               Yes           Electronic check   
7030  Month-to-month               Yes               Mailed check   
7031        Two year               Yes  Bank transfer (automatic)   

      monthly_charges  total_charges churn  
0               29.85          29.85    No  
1               56.95        1889.50    No  
2               53.85         108.15   Yes  
3               42.30        1840.75    No  
4               70.70         151.65   Yes  
...               ...            ...   ...  
7027            84.80        1990.50    No  
7028           103.20        7362.90    No  
7029            29.60         346.45    No  
7030            74.40         306.60   Yes  
7031           105.65        6844.50    No  

[7032 rows x 20 columns]

13. Let's check if everything is good now just to make sure. Check for Nan, Dtypes and duplicates

In [150]:
telecom_data_cleaned.isnull().sum()

gender               0
senior_citizen       0
partner              0
dependents           0
tenure               0
phone_service        0
multiple_lines       0
internet_service     0
online_security      0
online_backup        0
device_protection    0
tech_support         0
streaming_t_v        0
streaming_movies     0
contract             0
paperless_billing    0
payment_method       0
monthly_charges      0
total_charges        0
churn                0
dtype: int64

In [152]:
telecom_data_cleaned.dtypes

gender                object
senior_citizen         int64
partner               object
dependents            object
tenure                 int64
phone_service         object
multiple_lines        object
internet_service      object
online_security       object
online_backup         object
device_protection     object
tech_support          object
streaming_t_v         object
streaming_movies      object
contract              object
paperless_billing     object
payment_method        object
monthly_charges      float64
total_charges        float64
churn                 object
dtype: object

In [180]:
telecom_data_cleaned.duplicated().sum()

22

14. Now there are duplicate values, let's remove them 

## **4.Exploratory Data Analysis**

## **References**

[1] Solutions, H. (2024, April 2). Churn Management Basics: How to reduce customer churn. Hitachi Solutions. https://global.hitachi-solutions.com/blog/reduce-customer-churn/#:~:text=Churn%20%E2%80%94%20also%20known%20as%20customer%20churn%2C%20customer%20attrition%2C%20and%20customer%20turnover%20%E2%80%94%20is%20what%20happens%20when%20a%20customer%20ceases%20to%20use%20your%20product%20or%20service%20and%20terminates%20their%20relationship%20with%20your%20company.

[2] El-Abidin, R. (2024, September 25). 50 Customer Retention Statistics to Know. Hubspot. https://blog.hubspot.com/service/statistics-on-customer-retention#:~:text=Acquiring%20new%20customers%20is%20five%20times%20more%20expensive%20than%20retaining%20existing%20customers.

[3] Team, I. (2024, March 21). Churn Rate: What it means, examples, and calculations. Investopedia. https://www.investopedia.com/terms/c/churnrate.asp#:~:text=Churn%20rate%20in%20business,in%20a%20given%20period.

[4] Dzou, C. (2024, August 5). How to predict churn early and increase customer retention. Gong. https://www.gong.io/blog/predicting-churn/#:~:text=Predicting%20customer%20churn%20is%20crucial%20for%20businesses%20because%20it%20allows%20them%20to%20take%20proactive%20measures%20to%20retain%20their%20customer%20base%20and%20improve%20their%20bottom%20line.%C2%A0

[5] Understanding Churn Prediction with Machine Learning | Churned. (n.d.). https://www.churned.io/knowledge-base/understanding-churn-prediction-with-machine-learning#:~:text=Churn%20prediction%20is%20crucial%20for%20businesses%20as%20it%20enables%20them%20to%20identify%20potential%20churn%20risks%20before%20they%20occur.

[6] Prabadevi, B., Shalini, R., & Kavitha, B. (2023). Customer churning analysis using machine learning algorithms. International Journal of Intelligent Networks, 4, 145–154. https://doi.org/10.1016/j.ijin.2023.05.005

[7] Understanding Churn Prediction with Machine Learning | Churned. (n.d.). https://www.churned.io/knowledge-base/understanding-churn-prediction-with-machine-learning#:~:text=Churn%20prediction%20is%20crucial%20for%20businesses%20as%20it%20enables%20them%20to%20identify%20potential%20churn%20risks%20before%20they%20occur.

[8] Wagh, S. K., Andhale, A. A., Wagh, K. S., Pansare, J. R., Ambadekar, S. P., & Gawande, S. (2023). Customer churn prediction in telecom sector using machine learning techniques. Results in Control and Optimization, 14, 100342. https://doi.org/10.1016/j.rico.2023.100342

[9] Ullah I, Raza B, Malik AK, Imran M, Islam SU, Kim SW. A churn prediction model using random forest: analysis of machine learning, techniques for churn prediction and factor identification in telecom sector. IEEE Access May 6/2019. Link: https://ieeexplore.ieee.org/abstract/document/8706988 

[10] Wagh, S. K., Andhale, A. A., Wagh, K. S., Pansare, J. R., Ambadekar, S. P., & Gawande, S. (2023). Customer churn prediction in telecom sector using machine learning techniques. Results in Control and Optimization, 14, 100342. https://doi.org/10.1016/j.rico.2023.100342

[11] Ahmed, A. A., & Maheswari, D. (2017). Churn prediction on huge telecom data using hybrid firefly based classification. Egyptian Informatics Journal, 18(3), 215–220. https://doi.org/10.1016/j.eij.2017.02.002

[12] Jain, H., Khunteta, A., & Srivastava, S. (2020). Churn Prediction in Telecommunication using Logistic Regression and Logit Boost. Procedia Computer Science, 167, 101–112. https://doi.org/10.1016/j.procs.2020.03.187

[13] Poudel, S. S., Pokharel, S., & Timilsina, M. (2024). Explaining customer churn prediction in telecom industry using tabular machine learning models. Machine Learning With Applications, 17, 100567. https://doi.org/10.1016/j.mlwa.2024.100567

[14] Telecom customers. (2024, September 3). Kaggle. https://www.kaggle.com/datasets/tarekmuhammed/telecom-customers/data#:~:text=This%20dataset%20contains,for%20customer%20departure.